### ISSUE TITLE
Uncaught TypeError: Cannot read property 'messages' of undefined

### ISSUE THREAD

**Issue by gorkarod** [ricardofbarros/linter-js-standard#214]
[Enter steps to reproduce:]

1. ...
2. ...

**Atom**: 1.22.1 x64
**Electron**: 1.6.15
**OS**: Mac OS X 10.11.6
**Thrown From**: [linter-js-standard](https://github.com/ricardofbarros/linter-js-standard) package 4.1.0

### Stack Trace

Uncaught TypeError: Cannot read property 'messages' of undefined

```
At /Users/hartmut/.atom/packages/linter-js-standard/lib/utils/linter.js:33

TypeError: Cannot read property 'messages' of undefined
    at Object.module.exports (/packages/linter-js-standard/lib/utils/linter.js:33:31)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)
```

### Commands

```
     -0:07.4.0 tree-view:show (atom-workspace.workspace.scrollbars-visible-when-scrolling)
     -0:04.5.0 core:save (input.hidden-input)
```

### Non-Core Packages

```
advanced-open-file 0.16.7 
aligner 1.2.4 
aligner-coffeescript 1.0.1 
aligner-javascript 1.3.0 
aligner-php 1.2.0 
aligner-stylus 1.1.0 
atom-beautify 0.30.9 
atom-html-preview 0.2.5 
autocomplete-css-with-stylus-support 1.0.0 
autocomplete-php 0.3.7 
autohide-tree-view 0.25.1 
busy-signal 1.4.3 
color-picker 2.2.5 
copy-as-rtf 0.9.3 
dash 1.7.1 
duplicate-line-or-selection 0.9.0 
elm 0.1.0 
elm-format 3.0.0 
elmjutsu 7.1.5 
emmet 2.4.3 
file-icons 2.1.14 
FreshCut 1.0.0 
git-time-machine 1.5.9 
highlight-selected 0.13.1 
Hydrogen 2.1.0 
hyperclick 0.1.5 
incremental-search 5.2.2 
intentions 1.1.5 
js-hyperclick 1.12.2 
jumpy 4.1.1 
language-apache 1.7.0 
language-elm 1.5.0 
language-mjml 1.10.0 
language-pug 0.0.21 
language-vue 0.23.1 
language-vue-component 0.5.0 
linter 2.2.0 
linter-coffeescript 1.0.0 
linter-elm-make 0.27.2 
linter-htmlhint 1.4.0 
linter-js-standard 4.1.0 
linter-jshint 3.1.6 
linter-jsonlint 1.3.0 
linter-mjml 1.2.0 
linter-php 1.5.1 
linter-ui-default 1.6.10 
maximize-panes 0.2.0 
minimap-cursorline 0.2.0 
minimap-find-and-replace 4.5.2 
minimap-git-diff 4.3.1 
minimap-hide 0.3.0 
minimap-highlight-selected 4.6.1 
minimap-linter 2.1.3 
minimap-pigments 0.2.2 
minimap-plus 4.29.2 
mjml-preview 1.4.0 
node-debugger 1.10.1 
pigments 0.40.2 
prettier-atom 0.41.0 
pretty-json 1.6.4 
preview 0.17.0 
project-manager 3.3.5 
savey-wavey 0.3.0 
sort-lines 0.18.0 
standard-formatter 2.8.0 
standardjs-snippets 2.6.2 
Stylus 3.1.1 
stylus-autocompile 0.6.1 
svgo 2.8.5 
teletype 0.2.3 
toggle-quotes 1.0.1 
vue-stylefmt 0.5.0 
zentabs 0.8.9 
```

**Comment by sonicdoe** [comment]
Thanks for reporting this error. Are you able to reproduce this or was this a one-time error?

**Comment by gorkarod** [comment]
Sorry for keeping my reporting so short. In the meantime I did some investigation: Interestingly I have issues with a particular file in a particular project. When I copy it to e.g. desktop the problem disappears. Very strange …

**Comment by gorkarod** [comment]
The file looks like this
```javascript
function main () {
  processMenuItems()
  insertMainMenuCover()
  handleLangMenu()
}

function handleLangMenu () {
  const langswitch = Array.from(document.getElementsByClassName('module-metamenu--langswitch'))[0]
  const langmenu = Array.from(document.getElementsByClassName('module-metamenu--langmenu'))[0]
  langswitch.addEventListener('click', function (event) {
    if (!hasClass(langmenu, '-expanded')) {
      addClass(langmenu, '-expanded')
    } else {
      removeClass(langmenu, '-expanded')
    }
  })
}

function processMenuItems () {
  let menuItems = Array.from(document.getElementsByClassName('module-mainmenu--level-0'))[0].childNodes

  // Zap separator items
  for (let i = 0; i < menuItems.length; i++) {
    let item = menuItems[i]
    if (hasClass(item, 'separator')) {
      item.parentNode.removeChild(item)
    }
  }

  // Process others
  for (let i = 0; i < menuItems.length; i++) {
    let item = menuItems[i]
    let anchor = item.childNodes[0]
    let ul = Array.from(item.childNodes).filter((item) => item.nodeName === 'UL')
    const title = anchor.textContent
    const id = item.className.slice(1)
    const itemTop = item.offsetTop
    // attach Click handler
    if (ul.length) {
      anchor.addEventListener('click', function (event) {
        event.preventDefault()
        event.stopPropagation()
        addClass(ul[0].parentNode, '-expanded')
        let cover = document.getElementsByClassName('module-mainmenu--cover')[0]
        addClass(cover, '-expanded')
        return false
      })
      let container = document.createElement('div')
      addClass(container, 'module-mainmenu--level-1-container')
      container.style.paddingTop = itemTop + 'px'
      wrapElement(ul[0], container)

      let backbutton = document.createElement('button')
      backbutton.innerHTML = 'zurück'
      backbutton.addEventListener('click', function (event) {
        removeClass(this.parentNode, '-expanded')
        let cover = document.getElementsByClassName('module-mainmenu--cover')[0]
        removeClass(cover, '-expanded')
      })
      // addClass(backbutton, 'module-mainmenu--level-1-backbutton')
      insertBefore(backbutton, ul[0])

      let submenuheader = document.createElement('h2')
      submenuheader.innerHTML = title
      addClass(submenuheader, '-' + id)
      insertBefore(submenuheader, ul[0])
    }
    // Add icon container, wrap text
    anchor.innerHTML = `<div class="module-mainmenu--icon-level-0 -${id}"></div><div class="module-mainmenu--link-level-0">${title}</div>`
  }

  // Show menu when ready
  document.getElementsByClassName('module-mainmenu--level-0')[0].style.visibility = 'visible'
}

function insertMainMenuCover () {
  const cover = document.createElement('div')
  addClass(cover, 'module-mainmenu--cover')
  insertBefore(cover, document.getElementsByClassName('module-mainmenu--level-0')[0])
}

function hasClass (el, className) {
  return el.classList ? el.classList.contains(className) : new RegExp('\\b' + className + '\\b').test(el.className)
}

function addClass (el, className) {
  if (el.classList) el.classList.add(className)
  else if (!hasClass(el, className)) el.className += ' ' + className
}

function removeClass (el, className) {
  if (el.classList) el.classList.remove(className)
  else el.className = el.className.replace(new RegExp('\\b' + className + '\\b', 'g'), '')
}

function insertBefore (el, referenceNode) {
  referenceNode.parentNode.insertBefore(el, referenceNode)
}

function wrapElement (toWrap, wrapper = document.createElement('div')) {
  toWrap.parentNode.appendChild(wrapper)
  return wrapper.appendChild(toWrap)
}

// ---------------------------------------------------------------------------

if (document.readyState !== 'loading') {
  main()
} else if (document.addEventListener) {
  document.addEventListener('DOMContentLoaded', main)
} else {
  document.attachEvent('onreadystatechange', function () {
    if (document.readyState === 'complete') main()
  })
}
```

**Comment by sonicdoe** [comment]
> Interestingly I have issues with a particular file in a particular project.

Do you have `standard` installed as a development dependency in that project? Does running `standard` from the command line work correctly (for that particular file)?

**Comment by gorkarod** [comment]
Yes, if I e.g. put an extra semicolon at the end of a line and run `node_modules/.bin/standard src/main/frontend/js/scripts.js` I get a warning as expected.

**Comment by sonicdoe** [comment]
Which  `standard` version do you have installed? Do you have any custom `standard` settings in your `package.json`?

Could you also try running the following command (after putting an extra semicolon at the end of a line)?

```
$ node --print "require('standard').lintTextSync(fs.readFileSync(path.resolve('src/main/frontend/js/scripts.js'), 'utf8'))"
```

This runs the linter using [`standard`’s Node.js API](https://standardjs.com/index.html#is-there-a-nodejs-api) (which linter-js-standard uses). It _shouldn’t_ really differ from the command-line interface but I’d be interested in the output just to be sure.

**Comment by hacknlove** [comment]
I have this problem when I prepend any .js file with a dot

foo.js works fine
.foo.js fails

It fails also if I prepend any parent folder with a dor

foo/bar.js works fine
.foo/bar.js fails

**Comment by sonicdoe** [comment]
@hacknlove @gorkarod Thanks for the info, I was now able to reproduce this error.

[`standard` ignores hidden files and folders by default](https://github.com/standard/standard-engine#ignoring-files). Since we’re passing `filename` to `lintText()` in [v4.1.0](https://github.com/ricardofbarros/linter-js-standard/blob/master/CHANGELOG.md#410-2017-11-19), `standard` ignores the file causing `results` to be empty.

I’ll take a closer look at fixing this soon. Any help would be appreciated though 🙂

**Comment by gorkarod** [comment]
Great! Thanks @sonicdoe!

**Comment by sonicdoe** [comment]
This error has been fixed in https://github.com/ricardofbarros/linter-js-standard/commit/f553b4997e779725e565befdf73a1891d5907ac9 which will be included in the next major release.

I have also reverted the original commit introducing this error (https://github.com/ricardofbarros/linter-js-standard/commit/2d719175a161ec28a2e5cc3822e81dc580e36d1a) and published [v4.2.0](https://github.com/ricardofbarros/linter-js-standard/blob/v4.x/CHANGELOG.md#420-2018-01-02). Therefore, using v4.2.0, you should no longer run into this issue.

**Comment by technology-studio-apiary** [comment]
@sonicdoe Hi, I don't really understand why you reverted given option. If standard ignores, given files, then they should be ignored. 

But  is there way to do at least trade-off to introduce settings which allows disables/enables sending file to standard engine. 

Reverting the commit introduces back original problem #212. (Guys we have spent lot of time to identify given issue).

**Comment by sonicdoe** [comment]
I agree that if `standard` ignores a file, linter-js-standard should too. I will add passing the `filename` to `lintText()` back in the next major release. The reason I’ve reverted it in v4.2.0 is that v4.1.0 introduced a breaking change in a minor version ([which goes against the Semantic Versioning spec](https://semver.org/#what-do-i-do-if-i-accidentally-release-a-backwards-incompatible-change-as-a-minor-version)).

**Comment by rostislav-simonik** [comment]
@sonicdoe ok, thanks for update.  Until 5.0.0 released, we manually stick with 4.1.0.

**Comment by sonicdoe** [comment]
v5.0.0 has now been released and includes this fix.
